File: tests/Csls.TestProcessHost/DebuggerFixtureValue.cs
namespace Csls.TestProcessHost;

/// <summary>
/// Provides stable instance fields for debugger expansion tests.
/// </summary>
internal sealed class DebuggerFixtureValue
{
    /// <summary>
    /// Initializes the stable debugger value.
    /// </summary>
    /// <param name="number">The numeric field value.</param>
    /// <param name="text">The textual field value.</param>
    /// <param name="evaluationSignalPath">The path written when a cancelable evaluation starts.</param>
    internal DebuggerFixtureValue(int number, string text, string evaluationSignalPath)
    {
        Number = number;
        Text = text;
        Pair = (number, text);
        EvaluationSignalPath = evaluationSignalPath;
    }

    /// <summary>
    /// Gets the numeric field value.
    /// </summary>
    public int Number;

    /// <summary>
    /// Gets the textual field value.
    /// </summary>
    public string Text;

    /// <summary>
    /// Gets the named tuple field used for declared-field metadata tests.
    /// </summary>
    public readonly (int Code, string Label) Pair;

    /// <summary>
    /// Gets the file used to prove that target-code evaluation began.
    /// </summary>
    public readonly string EvaluationSignalPath;

    /// <summary>
    /// Gets the numeric field through a side-effect-free expression-bodied getter.
    /// </summary>
    public int NumberProperty => Number;

    /// <summary>
    /// Gets the numeric field through the compiler's debug return-local pattern.
    /// </summary>
    public int BlockNumberProperty
    {
        get
        {
            return Number;
        }
    }

    /// <summary>
    /// Gets the existing string reference through a side-effect-free getter.
    /// </summary>
    public string TextProperty => Text;

    /// <summary>
    /// Tests whether the getter result retains the exact original string object.
    /// </summary>
    /// <param name="value">The string supplied through the debugger's getter expression.</param>
    /// <returns>True when the supplied reference is the original string field.</returns>
    public bool IsOriginalText(string value) => ReferenceEquals(Text, value);

    /// <summary>
    /// Gets the tuple field through its declared element names.
    /// </summary>
    public (int Code, string Label) PairProperty => Pair;

    /// <summary>
    /// Gets the tuple through element names that differ from its backing field.
    /// </summary>
    public (int Number, string Text) RenamedPairProperty => Pair;

    /// <summary>
    /// Gets the tuple through an unnamed declaration.
    /// </summary>
    public (int, string) UnnamedPairProperty => Pair;

    /// <summary>
    /// Increments the numeric field before returning its changed value.
    /// </summary>
    public int MutatingNumberProperty => ++Number;

    /// <summary>
    /// Returns this object after recording one getter execution in its numeric field.
    /// </summary>
    public DebuggerFixtureValue MutatingSelfProperty
    {
        get
        {
            Number++;
            return this;
        }
    }

    /// <summary>
    /// Computes a tuple with property-specific names through target execution.
    /// </summary>
    public (int Next, string Caption) ComputedPairProperty => (Number + 1, Text);

    /// <summary>
    /// Throws from a property getter to verify supervised evaluation recovery.
    /// </summary>
    public int ThrowingNumberProperty => ThrowForDebugger();

    /// <summary>
    /// Signals property execution and waits for cooperative debugger cancellation.
    /// </summary>
    public int CancelableNumberProperty => WaitForDebuggerCancellation();

    /// <summary>
    /// Returns the original numeric value and mutates its field in a finally handler.
    /// </summary>
    public int FinallyNumberProperty
    {
        get
        {
            try
            {
                return Number;
            }
            finally
            {
                Number++;
            }
        }
    }

    /// <summary>
    /// Replaces the numeric field through ordinary target code when a fixture requires it.
    /// </summary>
    /// <param name="value">The replacement numeric value.</param>
    internal void SetNumber(int value) => Number = value;

    /// <summary>
    /// Replaces the textual field through ordinary target code when a fixture requires it.
    /// </summary>
    /// <param name="value">The replacement textual value.</param>
    internal void SetText(string value) => Text = value;

    /// <summary>
    /// Computes a stable value by executing target code for debugger evaluation tests.
    /// </summary>
    /// <returns>The numeric field incremented by one.</returns>
    internal int NextNumber() => Number + 1;

    /// <summary>
    /// Adds one debugger-supplied primitive argument to the stored number.
    /// </summary>
    /// <param name="value">The value supplied by managed function evaluation.</param>
    /// <returns>The stored number plus the supplied value.</returns>
    internal int AddForDebugger(int value) => Number + value;

    /// <summary>
    /// Selects the Int32 overload for exact debugger argument binding.
    /// </summary>
    /// <param name="value">The Int32 value supplied by managed function evaluation.</param>
    /// <returns>The stored number plus the supplied value.</returns>
    internal int AddOverloadedForDebugger(int value) => Number + value;

    /// <summary>
    /// Selects the Int64 overload for exact debugger argument binding.
    /// </summary>
    /// <param name="value">The Int64 value supplied by managed function evaluation.</param>
    /// <returns>The stored number plus the supplied value and an overload marker.</returns>
    internal long AddOverloadedForDebugger(long value) => Number + value + 100;

    /// <summary>
    /// Returns the length of a debugger-supplied string argument.
    /// </summary>
    /// <param name="value">The string supplied by managed function evaluation.</param>
    /// <returns>The supplied string length.</returns>
    internal int LengthForDebugger(string value) => value.Length + Number - 42;

    /// <summary>
    /// Returns the combined length of two debugger-materialized strings.
    /// </summary>
    /// <param name="first">The first string supplied by managed function evaluation.</param>
    /// <param name="second">The second string supplied by managed function evaluation.</param>
    /// <returns>The combined length of both strings.</returns>
    internal int CombinedLengthForDebugger(string first, string second) =>
        first.Length + second.Length + Number - 42;

    /// <summary>
    /// Tests whether a debugger-supplied reference is this receiver.
    /// </summary>
    /// <param name="value">The runtime reference supplied by managed function evaluation.</param>
    /// <returns>True when the supplied reference is this receiver.</returns>
    internal bool IsSameForDebugger(DebuggerFixtureValue? value) =>
        ReferenceEquals(this, value);

    /// <summary>
    /// Tests whether a debugger-supplied reference is null.
    /// </summary>
    /// <param name="value">The reference supplied by managed function evaluation.</param>
    /// <returns>True when the supplied reference is null.</returns>
    internal bool IsNullForDebugger(object? value)
    {
        GC.KeepAlive(Number);
        return value is null;
    }

    /// <summary>
    /// Throws a stable exception for debugger function-evaluation recovery tests.
    /// </summary>
    /// <returns>This method does not return normally.</returns>
    internal int ThrowForDebugger() =>
        throw new InvalidOperationException(Text);

    /// <summary>
    /// Runs until CoreCLR cooperatively aborts the debugger-owned evaluation.
    /// </summary>
    /// <returns>This method does not return normally.</returns>
    internal int WaitForDebuggerCancellation()
    {
        File.WriteAllText(EvaluationSignalPath, "started");
        while (true)
        {
            Thread.SpinWait(10_000);
        }
    }

    /// <summary>
    /// Exposes cooperative evaluation abort through files before completing the target's finally handler.
    /// </summary>
    /// <returns>This method exits through the debugger's cooperative abort.</returns>
    internal int WaitForDebuggerAbortRelease()
    {
        try
        {
            return WaitForDebuggerCancellation();
        }
        finally
        {
            File.WriteAllText(EvaluationSignalPath + ".aborting", "aborting");
            while (!File.Exists(EvaluationSignalPath + ".release"))
            {
                Thread.SpinWait(10_000);
            }
        }
    }

    /// <summary>
    /// Returns after the debugger test releases an evaluation through a real file boundary.
    /// </summary>
    /// <returns>The retained numeric value.</returns>
    internal int WaitForDebuggerRelease()
    {
        File.WriteAllText(EvaluationSignalPath, "started");
        while (!File.Exists(EvaluationSignalPath + ".release"))
        {
            Thread.SpinWait(10_000);
        }

        return Number;
    }
}

File: test-assets/Csls.Debugger.Fixtures.VisualBasic/DebuggerFixtureValue.vb
Imports System.Diagnostics

Namespace Global.Csls.Debugger.Fixtures.VisualBasic
    ''' <summary>
    ''' Provides a stable Visual Basic receiver for debugger function-evaluation tests.
    ''' </summary>
    <DebuggerDisplay("visual-basic={Me._number}", Type:="visual-basic-display")>
    Friend NotInheritable Class DebuggerFixtureValue
        ''' <summary>
        ''' Holds mutable static storage for debugger expression tests.
        ''' </summary>
        Friend Shared s_number As Integer

        Private ReadOnly _number As Integer

        ''' <summary>
        ''' Initializes the Visual Basic debugger receiver.
        ''' </summary>
        ''' <param name="number">The value returned by the debugger-visible method.</param>
        Friend Sub New(number As Integer)
            _number = number
        End Sub

        ''' <summary>
        ''' Gets the value returned by the debugger-visible method.
        ''' </summary>
        Friend ReadOnly Property Number As Integer
            Get
                Return _number
            End Get
        End Property

        ''' <summary>
        ''' Observes static storage through code compiled into the target process.
        ''' </summary>
        Friend Shared Function ReadStaticNumber() As Integer
            Return s_number
        End Function

        ''' <summary>
        ''' Initializes shared storage before the fixture stops for inspection.
        ''' </summary>
        ''' <param name="number">The input supplied to the fixture entry point.</param>
        Friend Shared Sub InitializeStaticNumber(number As Integer)
            s_number = number + 20
        End Sub

        ''' <summary>
        ''' Applies the target compiler's reference-null semantics to one value.
        ''' </summary>
        ''' <param name="value">The value tested by compiler-authored target code.</param>
        ''' <returns>True when the value is Nothing.</returns>
        Friend Shared Function CompilerIsNothing(value As Object) As Boolean
            Return value Is Nothing
        End Function

        ''' <summary>
        ''' Computes a stable result by executing target code.
        ''' </summary>
        ''' <returns>The stored number incremented by one.</returns>
        Friend Function NextNumber() As Integer
            Return Number + 1
        End Function

        ''' <summary>
        ''' Adds one debugger-supplied argument to the stored number.
        ''' </summary>
        ''' <param name="value">The value supplied by managed function evaluation.</param>
        ''' <returns>The stored number plus the supplied value.</returns>
        Friend Function AddNumber(value As Integer) As Integer
            Return Number + value
        End Function

        ''' <summary>
        ''' Combines named instance arguments with this receiver's stored number.
        ''' </summary>
        ''' <param name="first">The high-order value.</param>
        ''' <param name="second">The low-order value.</param>
        ''' <returns>The order-sensitive instance result.</returns>
        Friend Function CombineNamed(first As Integer, second As Integer) As Integer
            Return first * 100 + second + _number
        End Function

        ''' <summary>
        ''' Combines named static arguments in CLR parameter order.
        ''' </summary>
        ''' <param name="first">The high-order value.</param>
        ''' <param name="second">The low-order value.</param>
        ''' <returns>The order-sensitive static result.</returns>
        Friend Shared Function CombineNamedStatic(first As Integer, second As Integer) As Integer
            Return first * 100 + second
        End Function

        ''' <summary>
        ''' Exposes a Visual Basic optional parameter through loaded method metadata.
        ''' </summary>
        ''' <param name="first">The required high-order value.</param>
        ''' <param name="second">The optional low-order value.</param>
        ''' <returns>The order-sensitive result.</returns>
        Friend Shared Function CombineOptionalStatic(first As Integer, Optional second As Integer = 42) As Integer
            Return first * 100 + second
        End Function

        ''' <summary>
        ''' Exposes a scaled negative Visual Basic decimal default.
        ''' </summary>
        ''' <param name="amount">The optional decimal value.</param>
        ''' <returns>One when the target receives its exact decimal representation.</returns>
        Friend Shared Function OptionalDecimalStatic(Optional amount As Decimal = -12.50D) As Integer
            Dim bits = Decimal.GetBits(amount)
            Return If(bits(0) = 1250 AndAlso bits(1) = 0 AndAlso bits(2) = 0 AndAlso
                bits(3) = &H80020000, 1, 0)
        End Function

        ''' <summary>
        ''' Exposes a Visual Basic date default through method metadata.
        ''' </summary>
        ''' <param name="when">The optional date value.</param>
        ''' <returns>One when the target receives its declared date.</returns>
        Friend Shared Function OptionalDateStatic(Optional [when] As DateTime = #1/1/2020#) As Integer
            Return If([when].Ticks = 637134336000000000L AndAlso
                [when].Kind = DateTimeKind.Unspecified, 1, 0)
        End Function

        ''' <summary>
        ''' Returns the length of a debugger-materialized string.
        ''' </summary>
        ''' <param name="value">The string supplied by managed function evaluation.</param>
        ''' <returns>The supplied string length.</returns>
        Friend Function StringLength(value As String) As Integer
            Return value.Length + Number - 41
        End Function
    End Class
End Namespace
